src/main/java/dan200/computercraft/client/util/DirectVertexBuffer.java

```diff
@@ -48,11 +51,85 @@ public void upload( int vertexCount, VertexFormat.Mode mode, VertexFormat format
         sequentialIndices = true;
     }
 
-    public void drawWithShader( Matrix4f modelView, Matrix4f projection, ShaderInstance shader, int indexCount )
+    public void drawWithShader( Matrix4f modelView, Matrix4f projection, ShaderInstance shader, int vertexCount, int baseVertex )
     {
-        this.indexCount = indexCount;
-        drawWithShader( modelView, projection, shader );
-        this.indexCount = actualIndexCount;
+        indexCount = mode.indexCount( vertexCount );
+        drawWithShaderBaseVertex( modelView, projection, shader, baseVertex );
+        indexCount = actualIndexCount;
+    }
+
+    private void drawWithShaderBaseVertex( Matrix4f matrix4f, Matrix4f matrix4f2, ShaderInstance shaderInstance, int baseVertex )
```

Comment:
I think it's probably worth making `DirectVertexBuffer` its own class rather than inheriting from `VertexBuffer`. I originally subclassed in an attempt to be compatible with anybody who mixed into `drawWithShader`[^1], but seeing as we're overriding everything now it doesn't make much sense.

Will mean we can drop some access transformers too!


[^1]: Optifine. Iris doesn't appear to.

Reply:
I did this and then went a little crazy in the last commit trying to tailor `DirectVertexBuffer` to our specific use case in order to remove as much redundant setup as possible. Maybe it's too much? Let me know what you think.

Reply:
:+1: Probably worth removing the access widener lines, but otherwise looks very good!

Reply:
Removed the Access wideners :+1: